### PR TITLE
feat(compaction): support filter data by delete-range

### PR DESCRIPTION
**Pull request by Little-Wallace**
Signed-off-by: Little-Wallace <bupt2013211450@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

part of https://github.com/risingwavelabs/risingwave/issues/5978

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
- https://github.com/risingwavelabs/risingwave/issues/5978

### CODE REVIEW

File: src/storage/src/hummock/compactor/delete_range_aggregator.rs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        }
+        if self.seek_idx >= self.tombstone_index.len() {
+            return false;
+        }
+        self.tombstone_index[self.seek_idx]
+            .start_user_key
+            .as_slice()
+            .le(user_key)
+            && self.tombstone_index[self.seek_idx].sequence >= epoch
+            && self.tombstone_index[self.seek_idx].sequence <= self.watermark
```
Comment (soundOfDestiny) [inline src/storage/src/hummock/compactor/delete_range_aggregator.rs:187]: incorrect.
Suppose there are two tombstones X and Y.
key.epoch < X.seq < watermark < Y.seq
However, part of Y is replaced by X in `iter`
Therefore this key will not be deleted by Y, thus this key will not be deleted.
However, this key should have been deleted by X.

Comment (Little-Wallace) [reply src/storage/src/hummock/compactor/delete_range_aggregator.rs:187]: I have comment this condition. Delay deletion is still correct.

Comment (Little-Wallace) [reply src/storage/src/hummock/compactor/delete_range_aggregator.rs:187]: see cases in unit-test `test_delete_range_aggregator`. There is a `TODO` comment.

File: src/storage/src/hummock/sstable/builder.rs [hunk 2]
```diff
@@ -225,11 +234,16 @@ impl<W: SstableWriter> SstableBuilder<W> {
...
         let smallest_key = self.block_metas[0].smallest_key.clone();
-        let largest_key = self.last_full_key.clone();
+        let mut largest_key = self.last_full_key.clone();
 
         self.build_block().await?;
         let meta_offset = self.writer.data_len() as u64;
         assert!(!smallest_key.is_empty());
+        for (_, end_user_key) in &self.range_tombstones {
+            if end_user_key.as_slice().gt(user_key(&largest_key)) {
+                largest_key = key_with_epoch(prev_key(end_user_key).to_vec(), 0);
```
Comment (soundOfDestiny) [inline src/storage/src/hummock/sstable/builder.rs:244]: what about smallest key?

Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/builder.rs:244]: 🤔 good catch. I also need to update smallest key. Let me fix it.

Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/builder.rs]: fixed.

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            let tombstone = DeleteRangeTombstone {
+                start_user_key: start_user_key.to_vec(),
+                end_user_key,
+                sequence,
+            };
+            self.delete_tombstones.push(tombstone);
+        }
+    }
+
+    pub fn sort(&mut self) {
```
Comment (soundOfDestiny) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:109]: Since it is not different from sort condition of tombstone, shall we name it `sort_by_start`?

**Comment by soundOfDestiny** [comment]
Is delete range a closed interval or an open interval?

**Comment by Little-Wallace** [comment]
> Is delete range a closed interval or an open interval?

The left of it is closed while the right of it is open

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+pub struct DeleteRangeAggregatorIterator<I: DeleteRangeIterator> {
+    inner: I,
+    end_user_key_index: BinaryHeap<DeleteRangeTombstone>,
+    epoch_index: BTreeSet<HummockEpoch>,
+    watermark: u64,
+}
+
+impl<I: DeleteRangeIterator> DeleteRangeAggregatorIterator<I> {
+    pub fn should_delete(&mut self, target_key: &[u8], epoch: HummockEpoch) -> bool {
```
Comment (soundOfDestiny) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:236]: this function is correct
I have just checked equal case in this function, and it seems also correct.
You can check equal case in other codes and difference of user key and full key in other codes in you spare time.

**Review by soundOfDestiny — APPROVED** [review]

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+    pub fn new(key_range: KeyRange, watermark: u64, gc_delete_keys: bool) -> Self {
+        Self {
+            key_range,
+            delete_tombstones: vec![],
+            watermark,
+            gc_delete_keys,
+        }
+    }
+
+    pub fn add_tombstone(&mut self, data: Vec<(Vec<u8>, Vec<u8>)>) {
```
Comment (Li0k) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:73]: some comments or example for `add_tombstone`

File: src/storage/src/hummock/sstable/builder.rs [hunk 6]
```diff
@@ -224,12 +231,26 @@ impl<W: SstableWriter> SstableBuilder<W> {
...
+            vec![]
+        } else {
+            self.block_metas[0].smallest_key.clone()
+        };
+        let mut largest_key = self.last_full_key.clone();
 
         self.build_block().await?;
         let meta_offset = self.writer.data_len() as u64;
-        assert!(!smallest_key.is_empty());
+        for (start_key, end_user_key) in &self.range_tombstones {
```
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/storage/src/hummock/sstable/builder.rs:243]: My guess is that we need the delete range epoch. Can we use a more structural way to store the epoch instead of embeding it in the start key? For example, have a DeletedRange struct:
```rust
struct DeletedRange {
  start_user_key: Vec<u8>,
  end_user_key: Vec<u8>,
  epoch: u64
}
```

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+                if split_start_user_key.gt(end_user_key.as_slice()) {
+                    continue;
+                }
+                if split_start_user_key.gt(start_user_key) {
+                    start_user_key = split_start_user_key;
+                }
+            }
+            if !self.key_range.right.is_empty() {
+                let split_end_user_key = user_key(&self.key_range.right);
+                if split_end_user_key.le(start_user_key) {
```
Comment (Li0k) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:88]: Is it duplicated?

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+                start_user_key: start_user_key.to_vec(),
+                end_user_key,
+                sequence,
+            };
+            self.delete_tombstones.push(tombstone);
+        }
+    }
+
+    pub fn sort(&mut self) {
+        self.delete_tombstones.sort_by(|a, b| {
```
Comment (Li0k) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:110]: Can we use a sorted struct which depends on DeleteRangeTombstone compare function

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+    watermark: u64,
+}
+
+impl<I: DeleteRangeIterator> DeleteRangeAggregatorIterator<I> {
+    pub fn should_delete(&mut self, target_key: &[u8], epoch: HummockEpoch) -> bool {
+        if epoch >= self.watermark {
+            return false;
+        }
+
+        // take the smallest end_user_key which would never covery the current key and remove them
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:241]: nits: covery -> cover

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+use std::sync::Arc;
+
+use risingwave_hummock_sdk::key::{get_epoch, key_with_epoch, user_key};
+use risingwave_hummock_sdk::key_range::KeyRange;
+use risingwave_hummock_sdk::HummockEpoch;
+
+#[derive(Clone)]
+pub struct DeleteRangeTombstone {
+    start_user_key: Vec<u8>,
+    end_user_key: Vec<u8>,
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:26]: start_user_key is inclusive and end_user_key is exclusive? Need more documentations.

File: src/storage/src/hummock/sstable/builder.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/storage/src/hummock/sstable/builder.rs:243]: Just realized that this is exactly DeleteRangeTombstone. Can we use it instead?

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 10 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:26]: Yes.

Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:26]: > start_user_key is inclusive and end_user_key is exclusive? Need more documentations.

Yes. I have just asked.
> > Is delete range a closed interval or an open interval?
> 
> The left of it is closed while the right of it is open

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 8 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:110]: > Can we use a sorted struct which depends on DeleteRangeTombstone compare function

`DeleteRangeTombstone` has own compare function.

File: src/storage/src/hummock/sstable/builder.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Li0k) [reply src/storage/src/hummock/sstable/builder.rs:243]: `start_key is a full key with epoch and the end_key is a user key?`

I have the same doubt about this

[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/builder.rs:243]: Fixed.

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 7 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs]: Yes. I have removed it.

File: src/storage/src/hummock/compactor/mod.rs [hunk 11]
```diff
@@ -585,13 +593,36 @@ impl Compactor {
                 continue;
             }
 
+            if is_new_user_key && sst_builder.need_seal_current() {
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/compactor/mod.rs]: It has been removed.

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 10 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs]: I have add comment for this

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 9 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs]: fxied.

File: src/storage/src/hummock/compactor/compactor_runner.rs [hunk 12]
```diff
@@ -85,18 +89,50 @@ impl CompactorRunner {
...
+            for table_info in &level.table_infos {
+                let key_range = KeyRange::from(table_info.key_range.as_ref().unwrap());
+                if !self.key_range.full_key_overlap(&key_range) {
+                    continue;
+                }
+                let table = self
+                    .compactor
+                    .context
+                    .sstable_store
+                    .sstable(table_info, &mut local_stats)
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/compactor/compactor_runner.rs:127]: Yes. We always keep meta-cache in compactor.

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs]: After it is build, we would pass it as an `Arc` pointer so that it would never be update again.

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+            watermark,
+            gc_delete_keys,
+        }
+    }
+
+    pub fn add_tombstone(&mut self, data: Vec<DeleteRangeTombstone>) {
+        for mut tombstone in data {
+            if !self.key_range.left.is_empty() {
+                let split_start_user_key = user_key(&self.key_range.left);
+                if split_start_user_key.gt(tombstone.end_user_key.as_slice()) {
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:77]: `split_start_user_key` is inclusive while `tombstone.end_user_key` is exclusive. Should this be `ge` instead of `gt`?

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+                    continue;
+                }
+                if split_start_user_key.gt(tombstone.start_user_key.as_slice()) {
+                    tombstone.start_user_key = split_start_user_key.to_vec();
+                }
+            }
+            if !self.key_range.right.is_empty() {
+                let split_end_user_key = user_key(&self.key_range.right);
+                if split_end_user_key.le(tombstone.start_user_key.as_slice()) {
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:86]: `split_end_user_key` is inclusive and `tombstone.start_user_key` is also inclusive. Should this be `lt` instead of `le`?

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+
+#[cfg(test)]
+mod tests {
+    use bytes::Bytes;
+    use risingwave_hummock_sdk::key::key_with_epoch;
+
+    use super::*;
+
+    #[test]
+    pub fn test_delete_range_aggregator() {
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:266]: I think we need a unit test for the edge cases:
- `start_user_key` in the DeleteRangeAggregator key_range == `tombstone.start_user_key`
- `start_user_key` in the DeleteRangeAggregator key_range == `tombstone.end_user_key`
- `end_user_key` in the DeleteRangeAggregator key_range == `tombstone.start_user_key`
- `end_user_key` in the DeleteRangeAggregator key_range == `tombstone.end_user_key`

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+            self.inner.next();
+        }
+
+        // There may be several epoch, we only care the largest one.
+        self.epoch_index
+            .last()
+            .map(|tombstone_epoch| *tombstone_epoch >= epoch)
+            .unwrap_or(false)
+    }
+}
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:256]: I think `should_delete` assumes that `target_key` is passed in key order during the lifetime of DeleteRangeAggregatorIterator. We should document this assumption.

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 14 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:86]: No, `split_end_user_key` is  exclusive

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 13 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:77]: yes.

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 15 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:266]: done

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 13 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs]: fixed

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 16 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:258]: 🤔  I have add some comment

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 14 above)
Comment (hzxa21) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:86]: I see. I thought the split key range has the same property (left and right closed) of the sst key range :(

**Review by hzxa21 — APPROVED** [review]
LGTM
